AIP-86 - Add async support for Notifiers

ramitkataria opened the pull request:
This makes the `BaseNotifier` awaitable and implements the necessary changes required for the notifiers to work in non-blocking mode (required for DeadlineAlerts). Since notifiers use hooks which may need to use the TaskSDK API if they will fetching a `Connection`, I've added async counterparts to relevant TaskSDK functions as well, while avoiding as much code duplication as I could.

I've included changes needed to `SlackWebhookNotifier` as an example and will add support for some other common ones soon.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ramitkataria committed:
AIP-86 - Add async support for Notifiers

This makes the `BaseNotifier` awaitable and implements the necessary
changes required for the notifiers to work in non-blocking mode
(required for DeadlineAlerts). Since notifiers use hooks which may need
to use the TaskSDK API if they will fetching a `Connection`, I've added
async counterparts to relevant TaskSDK functions as well, while avoiding
as much code duplication as I could.

I've included changes needed to `SlackWebhookNotifier` as an example
and will add support for some other common ones soon.

ramitkataria committed:
Docs and naming fixes

ramitkataria committed:
CI fixes

o-nikolas commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -49,7 +49,21 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
             callback = import_string(self.callback_path)
-            result = await callback(**self.callback_kwargs)
+
+            # TODO: get airflow context
```
Is this coming in this PR or a future PR?

ramitkataria replied:
Likely a future PR

o-nikolas reviewed (commented):
I wonder if @ashb or @amoghrajesh want to have a look at the TaskSDK pieces here.

ramitkataria committed:
Make context a kwarg

ashb commented on providers/slack/tests/unit/slack/hooks/test_slack_webhook.py:
```diff
@@ -432,6 +432,7 @@ def test_hook_send_dict_legacy_slack_integration(self, mock_webhook_client_cls,
             {"text": "Test Text"},
             {"text": "Fallback Text", "blocks": ["Dummy Block"]},
             {"text": "Fallback Text", "blocks": ["Dummy Block"], "unfurl_media": True, "unfurl_links": True},
+            {"legacy": "value"},
```
?  Needed change or not?

ramitkataria replied:
I noticed that it supports other kwargs for legacy integration: https://github.com/apache/airflow/blob/85de1fe9b80e82db3f7b668bc23f56c66b1f75fc/providers/slack/src/airflow/providers/slack/hooks/slack_webhook.py#L246-L247
But that wasn't being tested in unit tests so I added this in

ashb commented on task-sdk/src/airflow/sdk/bases/hook.py:
```diff
@@ -56,21 +64,36 @@ def get_connection(cls, conn_id: str) -> Connection:
         return conn
 
+    @classmethod
+    async def async_get_connection(cls, conn_id: str) -> Connection:
```
I wonder if this should be called `aget_connection` instead.

I think this might be the pattern we've copied from somewhere else, but I can't think where right now... 🤔

ferruzzi replied:
I don't have an opinion either way on this one.  I searched the codebase and neither `aget_connection` nor `async_get_connection` seem to be in use at the moment, but there are examples of `aget_response` in the `trigger_job_runner`, and `aget_abstract` in a couple packages in the `site-packages/svcs` path, so it could make sense.

ramitkataria replied:
I came across `asend` as well. I've now changed this to `aget_connection`

ferruzzi replied:
By the same logic, should the new `async_notify` be `anotify`?

ramitkataria replied:
So far, I've only seen it used for `get` and `send` so I'm not sure if `anotify` would be obvious enough 🤔

ferruzzi replied:
I feel like it's not as intuitive in notify, and would argue (but not very strongly) for keeping it as `async_notify`, just thought I'd ask in case we wanted to standardize all of them

ashb commented on task-sdk/src/airflow/sdk/bases/hook.py:
```diff
@@ -56,21 +64,36 @@ def get_connection(cls, conn_id: str) -> Connection:
+        :return: connection
+        """
+        if not cls._is_task_sdk_context():
+            raise AirflowException("This method currently only supports getting connections through Task SDK")
```
```suggestion
            raise RuntimeException("This method currently only supports getting connections through Task SDK")
```

AirflowException is over used and has no special meaning, so I think RuntimeException is better suited here.

ramitkataria replied:
Thanks! Fixed now

ashb commented on task-sdk/src/airflow/sdk/bases/notifier.py:
```diff
@@ -104,4 +132,23 @@ def __call__(self, *args) -> None:
+        self._update_context(self.context)
+        self.render_template_fields(self.context)
+        try:
+            return self.async_notify(self.context).__await__()
```
Two questions here can't we make this a normal async function?

```suggestion
    async def __await__(self) -> Any:
        """
        Make the notifier awaitable.

        Context provided in the constructor is used.
        """
        self._update_context(self.context)
        self.render_template_fields(self.context)
        try:
            return await self.async_notify(self.context)
```

And why do we need the empty generator? I think with the above change that isn't needed anymore?

ramitkataria replied:
1. I'm quite certain `__await__` cannot be async. I tried and got a TypeError saying `__await__() returned a coroutine`
2. I added the empty generator to fix an error (`__await__() returned non-iterator of type NoneType`) when an exception is raised from `self.async_notify`. Later, I realized that I could've just used `yield`. Anyway, I changed the error handling so it lets the exception propagate so that the caller knows that there was a failure. So the `yield` isn't  needed anymore

ashb replied:
Oh yes. Then how about move all the logic (rendering, error handling) into an async function, and then this fn becomes:

```python
   async def _something(self):
        self._update_context(self.context)
        self.render_template_fields(self.context)
        ... # etc.

    def __await__(self):
        return self._something().__await__()
```
I think that would be cleanrer than heeding to have the empty generator.

ashb replied:
Or don't support await on the object, and have the api be `await n.async_notify()` directly. Either or.

ramitkataria replied:
After the error handling change I mentioned, we don't need the empty generator which makes it shorter than `__call__`. So I don't think we need to move it to a separate function. But if you would still like me to do it, then I think it should done for both `__call__` and `__await__`.
Also, I would like to keep the usage as similar to the current usage as possible so I want to avoid making the API `await n.async_notify()`

ashb replied:
If you've cleaned it up that's cool -- my major complaint was the exception handling and having to have the odd-to-me empyt generator

amoghrajesh commented on task-sdk/src/airflow/sdk/bases/hook.py:
```diff
@@ -56,21 +64,36 @@ def get_connection(cls, conn_id: str) -> Connection:
-
-        # if SUPERVISOR_COMMS is set, we're in task sdk context
-        if hasattr(sys.modules.get("airflow.sdk.execution_time.task_runner"), "SUPERVISOR_COMMS"):
+        if cls._is_task_sdk_context():
```
BTW, this is a temporary layer that going to go off for 3.1 through https://github.com/apache/airflow/issues/52520, so I do not think we should introduce a function for it.

ramitkataria replied:
I see, I removed the function

amoghrajesh commented on task-sdk/src/airflow/sdk/bases/hook.py:
```diff
@@ -56,21 +64,36 @@ def get_connection(cls, conn_id: str) -> Connection:
         return conn
 
+    @classmethod
+    async def async_get_connection(cls, conn_id: str) -> Connection:
```
Would be in favour of `aget_connection` too

ramitkataria committed:
PR changes

ramitkataria committed:
Merge branch 'main' of https://github.com/apache/airflow into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Exception handling in BaseNotifier

ramitkataria committed:
Update docstrings and typechecks

ferruzzi committed:
Update Notifiers docs pages

ferruzzi committed:
Add unit tests

ferruzzi committed:
Add Slack notifier unit tests

ferruzzi committed:
Add task-sdk unit tests

ferruzzi committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ferruzzi committed:
fix unit tests broken by the rebase

ferruzzi committed:
roll back changes to one unit test

This one actually made more sense how it was, asserting 0 first then adding one.

ferruzzi committed:
un-re-revert to remove pointless check in unit test

ramitkataria committed:
Refactor Deadline model unit tests to not make assertions about specific autogenerated IDs

ramitkataria committed:
Merge branch 'main' of github.com:aws-mwaa/upstream-to-airflow into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ferruzzi reviewed (approved):
I think we're good?

ramitkataria committed:
Re-add accidentally removed code comment about TaskSDK context

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ashb commented on airflow-core/docs/administration-and-deployment/logging-monitoring/callbacks.rst:
```diff
@@ -95,22 +101,43 @@ Before each task begins to execute, the ``task_execute_callback`` function will
+Using Notifiers
+^^^^^^^^^^^^^^^
+
+You can use Notifiers in your ``DAG`` definition by passing it as an argument to the ``on_*_callbacks``.
```
```suggestion
You can use Notifiers in your Dag definition by passing it as an argument to the ``on_*_callbacks``.
```
I think is the casing we are using now? Or if not that, at least don't put it in a code block

ferruzzi replied:
Fair, I'll make the change throughout that file.

ashb commented on airflow-core/src/airflow/triggers/deadline.py:
```diff
@@ -49,7 +49,21 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
+                callback_instance = callback(**self.callback_kwargs)
+
+                # Since parameters cannot be passed in __await__, context is passed as an attribute
+                callback_instance.context = context
```
Was there a reason why we set it as an attribute, instead of passing it in as another kwarg on L58?

```suggestion
                callback_instance = callback(**self.callback_kwargs, context=context)
```

Not every class is writable/can have attributes added to it. I'm not sure which approach is actually more "developer friendly".

ashb replied:
Do you have an example of what the callback class looks like that hits this code path?

ramitkataria replied:
Earlier, I did it that way to keep the constructor consistent across async and sync usage but I just realized that this code can be made simpler if I just pass it in the constructor so I've now added it as an optional parameter to the `BaseNotifier`. 

Notifiers are also the major examples of what these callback classes would look like in almost all cases but technically, it is possible to use any class that implements `__await__` and has a constructor that accepts context

ashb commented on airflow-core/tests/unit/models/test_deadline.py:
```diff
@@ -87,19 +98,7 @@ def setup_method():
-
-        assert session.query(Deadline).count() == 1
-
+    def test_add_deadline(self, dagrun, deadline_orm, session):
```
Is this test worth keeping? Does it actually test anything useful?

ferruzzi replied:
I think it is still verifying serialization/deserialization, isn't it? Create the Deadline object, write it to the db, read it back and make sure it's all still what we expected?

ashb replied:
That sounds very much like we are testing SQLA here. We should assume it has it's own unit tests

ashb commented on airflow-core/tests/unit/models/test_deadline.py:
```diff
@@ -142,32 +149,18 @@ def test_prune_deadlines(self, mock_session, conditions):
-            dagrun_id=RUN_ID,
-        )
-
+    def test_orm(self, deadline_orm, dagrun):
```
Ditto here -- this doesn't seem to be really testing anything anymore.

ferruzzi replied:
But I do agree that this one is pretty useless now.  I'll remove this one and we can chat about the other one.

ashb commented on task-sdk/src/airflow/sdk/bases/notifier.py:
```diff
@@ -33,11 +32,31 @@
+
+    Currently, the DAG/Task state change callbacks run on the DAG Processor and only support sync usage.
+
+    Usage:
```
```suggestion
    Usage::
    
```

(Then it'll get rendered as a code block if these docs ever rendered by sphinx)

ferruzzi replied:
I didn't realize that would work.  Neat.  I'll make the change and et back to you on the rest ASAP.  Thanks for the review.

ashb reviewed (commented):
Few minor Qs

ferruzzi committed:
Revisions for Ash

ramitkataria committed:
Pass context in async notifier constructor

ferruzzi committed:
Remove redundant unit test

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Add missing kwargs to parent class constructor

o-nikolas commented on airflow-core/tests/unit/triggers/test_deadline.py:
```diff
@@ -22,13 +22,29 @@
+    async def async_notify(self, context):
+        return f"Async notification: {self.message}, context: {context}"
+
+    def notify(self, context):
+        return f"Sync notification: {self.message}, context: {context}"
```
Do all Async Notifiers also have `sync(...)`?

ramitkataria replied:
Currently, any notifier either supports both async and sync, or only sync. But, it is not necessary to support both in case someone wants to use a notifier only for deadline async callbacks. I kept the implementation of sync notify in this case as an example of what a notifier should implement ideally so it can be used for both

o-nikolas reviewed (approved):
There is a lot to review here, but it seems reasonable 👍

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers

ramitkataria committed:
Merge branch 'main' into ramitkataria/deadlines/async-notifiers
